File: templates/python/code-carl/ch06_tree/t10_lc0222_count_nodes.py
"""LeetCode 222. Count Complete Tree Nodes
https://leetcode.cn/problems/count-complete-tree-nodes/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    root = TreeNode(
        1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), None)
    )
    solution = Solution()
    assert solution.countNodes(root) == 6
    solution = Solution()
    assert solution.countNodes(None) == 0
    print("All tests passed.")

File: solutions/code-carl/ch06_tree/T10_LC0222_CountNodes.java
/**
 * Practice template for:
 * - LeetCode 222. Count Complete Tree Nodes
 *   https://leetcode.cn/problems/count-complete-tree-nodes/
 */
public class T10_LC0222_CountNodes {
    public int countNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int h = 0;
        TreeNode left = root, right = root;
        while (left != null && right != null) {
            left = left.left;
            right = right.right;
            h++;
        }
        if (left == null && right == null) {
            return (int) Math.pow(2, h) - 1;
        }
        return countNodes(root.left) + countNodes(root.right) + 1;
    }
}
